File: solve.php
<?php
error_reporting(0);

define('V_NUM', 5000);
define('START_V', 1);
define('GOAL_V', V_NUM);

$Vs = array_fill(1, V_NUM, null);

function dfs($tmpVs, $tmpEs){

    // 頂点START_VからSTART_Vにはコスト0でいけるのでそれを記録し、$stackに入れる
    $tmpVs[START_V] = 0;
    $stack = new SplStack();
    $stack[] = array(0, START_V);

    // 調べるべき状態がもう無い($stackが空)なら探索終了
    while(!$stack->isEmpty()){
        // $nowVにたどり着くのに$nowTotalだけかかった状態から探索を進める
        $state = $stack->pop();
        $nowTotal = $state[0];
        $nowV = $state[1];

        // もうすでに$nowVに$nowTotalよりも小さいコストでたどり着けることが
        // わかっている場合は、この状態からの探索をやめる
        if(isset($tmpVs[$nowV]) && $tmpVs[$nowV] < $nowTotal) continue;
        
        // $nowVから出ている辺を全部みる
        foreach($tmpEs[$nowV] as $edge){
            $to = $edge[0];
            $cost = $edge[1];

            // 辺の先にある頂点にこれまでにたどり着いたことがない、
            // もしくはこれまでよりも小さいコストでたどり着けることがわかったら更新
            if(!isset($tmpVs[$to]) || $tmpVs[$to] > $tmpVs[$nowV] + $cost){
                $tmpVs[$to] = $tmpVs[$nowV] + $cost;

                // 探索を進めるべき状態として$stackに入れる
                $stack[]= array($tmpVs[$to], $to);
            }
        }
    }
    return $tmpVs[GOAL_V];
}

function bfs($tmpVs, $tmpEs){

    // 頂点START_Vから頂点START_Vにはコスト0でいけるのでそれを記録し、$queueに入れる
    $tmpVs[START_V] = 0;
    $queue = new SplQueue();
    $queue[] = array(0, START_V);

    // 調べるべき状態がもう無い($queueが空)なら探索終了
    while(!$queue->isEmpty()){

        // $nowVにたどり着くのに$nowTotalだけかかった状態から探索を進める
        $state = $queue->dequeue();
        $nowTotal = $state[0];
        $nowV = $state[1];

        // もうすでに$nowVに$nowTotalよりも小さいコストでたどり着けることが
        // わかっている場合は、この状態からの探索をやめる
        if(isset($tmpVs[$nowV]) && $tmpVs[$nowV] < $nowTotal) continue;

        // $nowVから出ている辺を全部みる
        foreach($tmpEs[$nowV] as $edge){
            $to = $edge[0];
            $cost = $edge[1];

            // 辺の先にある頂点にこれまでにたどり着いたことがない、
            // もしくはこれまでよりも小さいコストでたどり着けることがわかったら更新
            if(!isset($tmpVs[$to]) || $tmpVs[$to] > $tmpVs[$nowV] + $cost){
                $tmpVs[$to] = $tmpVs[$nowV] + $cost;

                // 探索を進めるべき状態として$queueに入れる
                $queue[]= array($tmpVs[$to], $to);
            }
        }
    }
    return $tmpVs[GOAL_V];
}

function dijkstra($tmpVs, $tmpEs){
    // 頂点START_Vから頂点START_Vにはコスト0でいけるのでそれを記録し、heapに入れる
    $tmpVs[START_V] = 0;
    $heap = new SplMinHeap();
    $heap->insert(array(0, START_V));

    // 調べるべき状態がもう無い(heapが空)なら探索終了
    while(!$heap->isEmpty()){

        // $nowVにたどり着くのに$nowTotalだけかかった状態から探索を進める
        $state = $heap->extract();
        $nowTotal = $state[0];
        $nowV = $state[1];

        // もうすでに$nowVに$nowTotalよりも小さいコストでたどり着けることが
        // わかっている場合は、この状態からの探索をやめる
        if(!isset($tmpVs[$nowV]) && $tmpVs[$nowV] < $nowTotal) continue;

        // $nowVから出ている辺を全部みる
        foreach($tmpEs[$nowV] as $edge){
            $to = $edge[0];
            $cost = $edge[1];

            // 辺の先にある頂点にこれまでにたどり着いたことがない、
            // もしくはこれまでよりも小さいコストでたどり着けることがわかったら更新
            if(!isset($tmpVs[$to]) || $tmpVs[$to] > $tmpVs[$nowV] + $cost){
                $tmpVs[$to] = $tmpVs[$nowV] + $cost;

                // 探索を進めるべき状態として$heapに入れる
                $heap->insert(array($tmpVs[$to], $to));
            }
        }
    }
    return $tmpVs[GOAL_V];
}

$csv = new SplFileObject(__DIR__ . '/GraphE.csv', 'r');
$csv->setFlags(SplFileObject::READ_CSV);
$cnt = 0;

foreach ($csv as $row) {
    if ($row === [null]) continue; // 最終行の処理
    $edge = array((int)$row[1], (int)$row[2]);
    $Edges[(int)$row[0]][] = $edge;
}

$tmpVs = $Vs;
$tmpEs = $Edges;

$startT = microtime(true);
$ret = dfs($tmpVs, $tmpEs);
$endT = microtime(true);
echo "DFS   result:" . $ret . " time:" . ($endT - $startT) . "秒" . PHP_EOL;


$tmpVs = $Vs;
$tmpEs = $Edges;

$startT = microtime(true);
$ret = bfs($tmpVs, $tmpEs);
$endT = microtime(true);
echo "BFS   result:" . $ret . " time:" . ($endT - $startT) . "秒" . PHP_EOL;


$tmpVs = $Vs;
$tmpEs = $Edges;

$startT = microtime(true);
$ret = dijkstra($tmpVs, $tmpEs);
$endT = microtime(true);
echo "Dijkstra   result:" . $ret . " time:" . ($endT - $startT) . "秒". PHP_EOL;

File: solve.cpp
#include<fstream>
#include<iostream>
#include<vector>
#include<string>
#include<sstream>
#include<queue>
#include<stack>
#include<functional>
#include<chrono>

using namespace std;
using namespace chrono;
using ll = long long;
using P  = pair<ll,int>;

struct edge{
    int to, cost;
    edge(const int& to, const int& cost = 1): to(to), cost(cost){}
    edge(const edge &e): to(e.to), cost(e.cost) {}
};

constexpr int V_NUM = 5000;
constexpr int START_V = 0;
constexpr int GOAL_V = V_NUM - 1;

vector<ll> V(V_NUM, -1);
vector<vector<edge>> Es(V_NUM);
vector<ll> tmpVs;
vector<vector<edge>> tmpEs;
stack<P> st;
queue<P> q;
priority_queue<P, vector<P>, greater<P>> pq;

ll dfs(vector<ll>& tmpVs, vector<vector<edge>>& tmpEs){
    ll ret = 0;
    tmpVs[START_V] = 0;
    st.emplace(P(0, START_V));
    while(!st.empty()){
        P state = st.top(); st.pop();
        ll nowTotal = state.first;
        int nowV = state.second;
        if (tmpVs[nowV] != -1 && tmpVs[nowV] < nowTotal) continue;
        for(int i = 0; i < tmpEs[nowV].size(); ++i){
            edge e = tmpEs[nowV][i];
            if(tmpVs[e.to] == -1 || tmpVs[e.to] > tmpVs[nowV] + e.cost){
                tmpVs[e.to] = tmpVs[nowV] + e.cost;
                st.emplace(P(tmpVs[e.to], e.to));
            }
        }
    }
    ret = tmpVs[GOAL_V];
    return ret;
}

ll bfs(vector<ll>& tmpVs, vector<vector<edge>>& tmpEs){
    ll ret = 0;
    tmpVs[START_V] = 0;
    q.emplace(P(0, START_V));
    while(!q.empty()){
        P state = q.front(); q.pop();
        ll nowTotal = state.first;
        int nowV = state.second;
        if (tmpVs[nowV] != -1 && tmpVs[nowV] < nowTotal) continue;
        for(int i = 0; i < tmpEs[nowV].size(); ++i){
            edge e = tmpEs[nowV][i];
            if(tmpVs[e.to] == -1 || tmpVs[e.to] > tmpVs[nowV] + e.cost){
                tmpVs[e.to] = tmpVs[nowV] + e.cost;
                q.emplace(P(tmpVs[e.to], e.to));
            }
        }
    }
    ret = tmpVs[GOAL_V];
    return ret;
}

ll dijkstra(vector<ll>& tmpVs, vector<vector<edge>>& tmpEs){
    ll ret = 0;
    tmpVs[START_V] = 0;
    pq.emplace(P(0, START_V));
    while (!pq.empty()) {
        P state = pq.top(); pq.pop();
        ll nowTotal = state.first;
        int nowV = state.second;
        if (tmpVs[nowV] != -1 && tmpVs[nowV] < nowTotal) continue;
        for (int i = 0; i < tmpEs[nowV].size(); i++) {
            edge e = tmpEs[nowV][i];
            if (tmpVs[e.to] == -1 || tmpVs[e.to] > tmpVs[nowV] + e.cost) {
                tmpVs[e.to] = tmpVs[nowV] + e.cost;
                pq.emplace(P(tmpVs[e.to], e.to));
            }
        }
    }
    ret = tmpVs[GOAL_V];
    return ret;
}

int main(int argc, char** argv)
{
    ifstream ifs("GraphE.csv");
    if(!ifs){
        cout<<"couldn't open file"<<endl;
        return 1;
    }

    string row;
    while(getline(ifs,row)){
        string token;
        istringstream istrstr(row);

        int e_data[3];int idx = 0;
        while(getline(istrstr, token, ',')){
            stringstream strstr;
            strstr << token;
            strstr >> e_data[idx++];
        }
        Es[e_data[0] - 1].emplace_back(edge(e_data[1] - 1, e_data[2]));
    }

    tmpVs = V;
    tmpEs = Es;
    
    auto duration = std::chrono::high_resolution_clock::now() - time_point;

auto count = std::chrono::duration_cast<std::chrono::microseconds>(duration).count();

    ll ret = 0;
    auto startT = system_clock::now();
    //auto startT = high_resolution_clock::now();
    ret = dfs(tmpVs, tmpEs);
    auto endT = system_clock::now();
    //auto endT = high_resolustion_clock::now();
    auto t = duration_cast<milliseconds>(endT - startT).count();
    //auto t = duration_cast<microseconds>(endT - startT).count();
    cout << "DFS   result:" << ret << " time:" << t << "ms" << endl;
    //cout << "DFS   result:" << ret << " time:" << t << "μs" << endl;
    
    tmpVs = V;
    tmpEs = Es;

    startT = system_clock::now();
    ret = bfs(tmpVs, tmpEs);
    endT = system_clock::now();
    t = duration_cast<milliseconds>(endT - startT).count();
    cout << "BFS   result:" << ret << " time:" << t << "ms" << endl;

    tmpVs = V;
    tmpEs = Es;

    startT = system_clock::now();
    ret = dijkstra(tmpVs, tmpEs);
    endT = system_clock::now();
    t = duration_cast<milliseconds>(endT - startT).count();
    cout << "Dijkstra   result:" << ret << " time:" << t << "ms" << endl;
    
    return 0;
}
